### PR TITLE
codygateway: Initial implementation of embeddings API

### PR DESCRIPTION
**Pull request by eseliger**
This PR adds support for a sourcegraph embeddings provider and backend routes on the Cody Gateway side for embeddings, plus rate limit configuration and UIs on the Sourcegraph side.

It shouldn't affect existing instances at all so we can roll this out and see how it performs on some instance. 

Closes https://github.com/sourcegraph/sourcegraph/issues/52644
Closes https://github.com/sourcegraph/sourcegraph/issues/52645

## Test plan

Manually tested on a bunch of embeddings jobs, observing the outbound request log.

### CODE REVIEW

File: migrations/frontend/squashed.sql [hunk 1]
```diff
@@ -3919,12 +3919,21 @@ CREATE TABLE product_subscriptions (
     cody_gateway_enabled boolean DEFAULT false NOT NULL,
     cody_gateway_chat_rate_limit integer,
     cody_gateway_chat_rate_interval_seconds integer,
+    cody_gateway_embeddings_api_rate_limit integer,
+    cody_gateway_embeddings_api_rate_interval_seconds integer,
+    cody_gateway_embeddings_api_allowed_models text[],
```
Comment (eseliger) [inline migrations/frontend/squashed.sql:3924]: not getting prettier.. 😆 Let's reduce the column count in a follow up, would rather want to not clutter this more and get something out there quickly so we can dogfood 😬

File: enterprise/cmd/cody-gateway/shared/config.go [hunk 2]
```diff
@@ -69,17 +71,19 @@ func (c *Config) Load() {
...
+	c.OpenAI.AllowedModels = splitMaybe(c.Get("CODY_GATEWAY_OPENAI_ALLOWED_MODELS", "gpt-4,gpt-3.5-turbo", "The models for chat completions allowed to be used with OpenAI."))
+
+	c.AllowedEmbeddingsModels = splitMaybe(c.Get("CODY_GATEWAY_ALLOWED_EMBEDDINGS_MODELS", "openai/text-embedding-ada-002", "The models allowed for embeddings generation."))
 
 	c.AllowAnonymous = c.GetBool("CODY_GATEWAY_ALLOW_ANONYMOUS", "false", "Allow anonymous access to Cody Gateway.")
 	c.SourcesSyncInterval = c.GetInterval("CODY_GATEWAY_SOURCES_SYNC_INTERVAL", "2m", "The interval at which to sync actor sources.")
 
 	c.BigQuery.ProjectID = c.Get("CODY_GATEWAY_BIGQUERY_PROJECT_ID", os.Getenv("GOOGLE_CLOUD_PROJECT"), "The project ID for the BigQuery events.")
-	c.BigQuery.Dataset = c.Get("CODY_GATEWAY_BIGQUERY_DATASET", "CODY_GATEWAY", "The dataset for the BigQuery events.")
+	c.BigQuery.Dataset = c.Get("CODY_GATEWAY_BIGQUERY_DATASET", "cody_gateway", "The dataset for the BigQuery events.")
```
Comment (eseliger) [inline enterprise/cmd/cody-gateway/shared/config.go:86]: I think this was a typo

File: cmd/frontend/graphqlbackend/dotcom.graphql [hunk 3]
```diff
@@ -377,6 +381,7 @@ type CodyGatewayRateLimit {
     allowedModels: [String!]!
     """
     Requests per time interval.
+    TODO: This needs to become BigInt, the limits for embeddings can exceed int32.
```
Comment (eseliger) [inline cmd/frontend/graphqlbackend/dotcom.graphql:384]: Filed https://github.com/sourcegraph/sourcegraph/issues/53007 as a follow up for this, so we can get this out earlier for testing.

File: enterprise/cmd/cody-gateway/internal/httpapi/embeddings/openai.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+	if resp.StatusCode >= 300 || resp.StatusCode < 200 {
+		// If we are being rate limited by OpenAI, we don't want to forward that error and instead
+		// return a 503 to the client. It's not them being limited, it's us and that an operations
+		// error on our side.
+		if resp.StatusCode == http.StatusTooManyRequests {
+			return nil, 0, response.NewHTTPStatusCodeError(http.StatusServiceUnavailable, errors.Newf("we're facing too much load at the moment, please retry"))
+		}
+		respBody, _ := io.ReadAll(io.LimitReader(resp.Body, 1024))
+		return nil, 0, response.NewHTTPStatusCodeError(resp.StatusCode, errors.Errorf("embeddings: %s %q: failed with status %d: %s", req.Method, req.URL.String(), resp.StatusCode, string(respBody)))
```
Comment (eseliger) [inline enterprise/cmd/cody-gateway/internal/httpapi/embeddings/openai.go:67]: q: is forwarding status codes what we want here?

File: enterprise/cmd/cody-gateway/internal/httpapi/embeddings/models.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	upstreamName string
+	dimensions   int
+}
+
+var openAIModelMappings = map[string]openAIModel{
+	"openai/text-embedding-ada-002": {
+		upstreamName: "text-embedding-ada-002",
+		dimensions:   1536,
+	},
+}
```
Comment (eseliger) [inline enterprise/cmd/cody-gateway/internal/httpapi/embeddings/models.go:43]: todo this needs some cleanup

File: client/web/src/enterprise/site-admin/dotcom/productSubscriptions/CodyServicesSection.tsx [hunk 6]
```diff
@@ -146,20 +151,50 @@ export const CodyServicesSection: React.FunctionComponent<Props> = ({
                                             productSubscriptionID={productSubscriptionID}
                                             rateLimit={codyGatewayAccess.chatCompletionsRateLimit}
                                             refetchSubscription={refetchSubscription}
-                                            title="Chat rate limit"
+                                            title="Chat and recipes"
```
Comment (bobheadxi) [inline client/web/src/enterprise/site-admin/dotcom/productSubscriptions/CodyServicesSection.tsx:154]: This is the only mention of `recipes` anywhere, is this change intentional?

File: enterprise/cmd/cody-gateway/internal/actor/actor.go [hunk 7]
```diff
@@ -40,8 +40,12 @@ type Actor struct {
...
 	// archived, if access is enabled, and if any rate limits are set.
 	AccessEnabled bool `json:"accessEnabled"`
-	// RateLimits holds the rate limits for Cody Gateway access for this actor.
-	RateLimits map[types.CompletionsFeature]RateLimit `json:"rateLimits"`
+	// CompletionsAccessEnabled holds the rate limits for Cody Gateway access to completions
+	// for this actor.
+	CompletionsRateLimits map[types.CompletionsFeature]RateLimit `json:"completionsRateLimits"`
+	// EmbeddingsRateLimit holds the rate limit for Cody Gateway access to embeddings
+	// for this actor.
+	EmbeddingsRateLimit RateLimit `json:"embeddingsRateLimit"`
```
Comment (bobheadxi) [inline enterprise/cmd/cody-gateway/internal/actor/actor.go:48]: any chance we might need a `map[types.EmbeddingsFeature]RateLimit` type thing? e.g. with different embeddings models, embeddings classes (pay extra for superrepo embeddings) etc?

File: enterprise/cmd/cody-gateway/internal/actor/actor.go [hunk 8]
```diff
@@ -93,28 +97,51 @@ func WithActor(ctx context.Context, a *Actor) context.Context {
 	return context.WithValue(ctx, actorKey, a)
 }
 
-func (a *Actor) Limiter(redis limiter.RedisStore, feature types.CompletionsFeature) (limiter.Limiter, bool) {
+func (a *Actor) CompletionsLimiter(redis limiter.RedisStore, feature types.CompletionsFeature) (limiter.Limiter, bool) {
```
Comment (bobheadxi) [inline enterprise/cmd/cody-gateway/internal/actor/actor.go:100]: feels a bit leaky for actor to have so many limiter methods - could we abstract over just `feature types.Feature` which covers all the completion types and `types.FeatureEmbeddings`?

File: client/web/src/enterprise/site-admin/dotcom/productSubscriptions/CodyServicesSection.tsx (hunk 6 above)
Comment (eseliger) [reply client/web/src/enterprise/site-admin/dotcom/productSubscriptions/CodyServicesSection.tsx:154]: The headline of this section is "Rate limits" and it felt redundant when looking at it.. no much thought has went into this "decision" 😀

File: enterprise/cmd/cody-gateway/internal/actor/actor.go (hunk 7 above)
Comment (eseliger) [reply enterprise/cmd/cody-gateway/internal/actor/actor.go:48]: maaybe, as with the move from chat only to chat and completions limiting, I expect that move to be fairly transparent should it be needed eventually, but this feels a little less like overengineering for a future where we might have very different requirements in 2 months from now anyways 😓 😬

File: enterprise/cmd/cody-gateway/internal/actor/actor.go (hunk 8 above)
Comment (eseliger) [reply enterprise/cmd/cody-gateway/internal/actor/actor.go:100]: I thought about that, too.. types.CompletionsFeature currently serves as the type used in the X-Sourcegraph-Feature header and adding a new value to it that isn't actually allowed through the API felt equally strange to me though :/ What do you think would look the cleanest here?

Comment (bobheadxi) [reply enterprise/cmd/cody-gateway/internal/actor/actor.go]: I think `types.Feature` would be ideal, and we can restructure e.g. `EmbeddingsRateLimit` around it too - I think it makes sense that only some endpoints support some features, and we reject requests to upstream handlers that use features that are unknown to them

for example anthropic upstream handler only knows about `chat` and `completions`, so we reject every other feature type as unsupported, while embeddings upstream handler only knows about `embeddings` perhaps?

File: enterprise/cmd/cody-gateway/internal/actor/actor.go (hunk 7 above)
Comment (bobheadxi) [reply enterprise/cmd/cody-gateway/internal/actor/actor.go:54]: https://github.com/sourcegraph/sourcegraph/pull/52810#discussion_r1220459622 would be better actually, where we have `map[types.Feature]RateLimit` - IMO this isn't overengineering much as it is saving us some repetition :)

File: enterprise/cmd/cody-gateway/internal/actor/actor.go (hunk 8 above)
Comment (eseliger) [reply enterprise/cmd/cody-gateway/internal/actor/actor.go]: aight, let me tinker with that :)

File: enterprise/cmd/cody-gateway/internal/actor/actor.go (hunk 7 above)
Comment (eseliger) [reply enterprise/cmd/cody-gateway/internal/actor/actor.go]: Implemented with the below idea of having a feature enum!

File: enterprise/cmd/cody-gateway/internal/actor/actor.go (hunk 8 above)
Comment (eseliger) [reply enterprise/cmd/cody-gateway/internal/actor/actor.go]: done!

File: enterprise/cmd/cody-gateway/internal/actor/actor.go [hunk 9]
```diff
@@ -125,6 +129,11 @@ func (a *Actor) Limiter(
 	if concurrencyLimit < 1 {
 		concurrencyLimit = 1
 	}
+	if feature == codygateway.FeatureEmbeddings {
+		// For embeddings, we use a custom, hardcoded limit. Concurrency on the client side
+		// should be 1, so 15 is a very safe default here.
+		concurrencyLimit = 15
+	}
```
Comment (bobheadxi) [inline enterprise/cmd/cody-gateway/internal/actor/actor.go:136]: should we have this condition further up to save us the `concurrencyLimit` arithmetic in this case?

File: enterprise/cmd/cody-gateway/internal/httpapi/embeddings/handler.go [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+			}
+		}()
+
+		resp, ut, err := c.GenerateEmbeddings(r.Context(), body)
+		usedTokens = ut
+		upstreamFinished = time.Since(upstreamStarted)
+		if err != nil {
+			logger.Error(
+				"error from upstream",
+				log.String("client", fmt.Sprintf("%T", c)),
```
Comment (bobheadxi) [inline enterprise/cmd/cody-gateway/internal/httpapi/embeddings/handler.go:102]: nit - create a logger when we evaluate `c` and add this field via `logger.With`, so that `response.JSONError` events will include it as well

File: enterprise/cmd/cody-gateway/internal/limiter/limiter.go [hunk 11]
```diff
@@ -14,7 +14,7 @@ type Limiter interface {
 	// actually consume a request. This allows us to easily avoid deducting from
 	// the rate limit if the request to upstream fails, at the cost of slight
 	// over-allowance.
-	TryAcquire(ctx context.Context) (commit func() error, err error)
+	TryAcquire(ctx context.Context) (commit func(int) error, err error)
```
Comment (bobheadxi) [inline enterprise/cmd/cody-gateway/internal/limiter/limiter.go:17]:
```suggestion
	//
	// The commit callback accepts a parameter that dictates how much rate
	// limit to consume for this request.
	TryAcquire(ctx context.Context) (commit func(int) error, err error)
```

File: enterprise/cmd/cody-gateway/internal/httpapi/embeddings/openai.go (hunk 4 above)
Comment (bobheadxi) [reply enterprise/cmd/cody-gateway/internal/httpapi/embeddings/openai.go:67]: Maybe we should never forward anything other than 2xx and 503 - unauthorized is also an operations error from our side for example

**Review by bobheadxi — APPROVED** [review]
most comments are minor, approving to unblock 😁 good stuff, let's test this out!
